src/components/tl/ucp/barrier/barrier_knomial.c

```diff
@@ -1,21 +1,150 @@
+    do {                                                                       \
+        iteration = task->barrier.iteration;                                   \
+        radix_pow = task->barrier.radix_mask_pow;                              \
+    } while (0)
+
+#define SAVE_STATE(_phase)                                                     \
+    do {                                                                       \
+        task->barrier.phase          = _phase;                                 \
+        task->barrier.iteration      = iteration;                              \
+        task->barrier.radix_mask_pow = radix_pow;                              \
+    } while (0)
 
 ucc_status_t ucc_tl_ucp_barrier_knomial_progress(ucc_coll_task_t *coll_task)
 {
     ucc_tl_ucp_task_t *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
-    /* TODO implement main logic */
+    ucc_tl_ucp_team_t *team       = task->team;
+    int                myrank     = team->rank;
+    int                group_size = team->size;
+    int                radix      = task->barrier.radix;
+    int full_tree_size, pow_k_sup, n_full_subtrees, full_size, node_type;
+    int iteration, radix_pow, k, step_size, peer;
+
+    KN_RECURSIVE_SETUP(radix, myrank, group_size, pow_k_sup, full_tree_size,
+                       n_full_subtrees, full_size, node_type);
+    RESTORE_STATE();
+    GOTO_PHASE(task->barrier.phase);
+
+    if (KN_NODE_EXTRA == node_type) {
+        peer = KN_RECURSIVE_GET_PROXY(myrank, full_size);
+        ucc_tl_ucp_send_nb(NULL, 0, UCC_MEMORY_TYPE_UNKNOWN, peer, team, task);
+        ucc_tl_ucp_recv_nb(NULL, 0, UCC_MEMORY_TYPE_UNKNOWN, peer, team, task);
+    }
+
+    if (KN_NODE_PROXY == node_type) {
+        peer = KN_RECURSIVE_GET_EXTRA(myrank, full_size);
+        ucc_tl_ucp_recv_nb(NULL, 0, UCC_MEMORY_TYPE_UNKNOWN, peer, team, task);
+    }
+PHASE_EXTRA:
+    if (KN_NODE_PROXY == node_type || KN_NODE_EXTRA == node_type) {
+        if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+            SAVE_STATE(PHASE_EXTRA);
+            return UCC_INPROGRESS;
+        }
+        if (KN_NODE_EXTRA == node_type) {
+            goto completion;
+        }
+    }
+
+    for (; iteration < pow_k_sup; iteration++) {
+        step_size = radix_pow * radix;
+        for (k = 1; k < radix; k++) {
+            peer = (myrank + k * radix_pow) % step_size +
+                   (myrank - myrank % step_size);
+            if (peer >= full_size)
+                continue;
+            ucc_tl_ucp_send_nb(NULL, 0, UCC_MEMORY_TYPE_UNKNOWN, peer, team,
+                               task);
+        }
+
+        for (k = 1; k < radix; k++) {
+            peer = (myrank + k * radix_pow) % step_size +
+                   (myrank - myrank % step_size);
+            if (peer >= full_size)
+                continue;
+            ucc_tl_ucp_recv_nb(NULL, 0, UCC_MEMORY_TYPE_UNKNOWN, peer, team,
+                               task);
+        }
+        radix_pow *= radix;
```

Comment:
It seems like all sends/recvs from all phases are on the same task, is that correct?

Reply:
correct. we touched upon this several times long ago. basically we defined tasks and schedules. From that point, it is up to CL/TL to put whatever logic they want into a TASK. In TL/UCP we put "collective primitives" into task - ie, complete implementation of a collective over a tl_ucp_team is a task. This way we don't need to generate long schedules, allocate multiple task objects and move all this through the progress queue engine. There is an alternative: build schedule and move algorithm steps via progress_queue. This would be "event_driven" approach, aka, ucg-like. It is just another approach with its pros/cons. This could be another TL implementation if at some point we will see benefit from this.

Reply:
Right. With the infrastructure, it seems very straightforward to write algorithms in both approaches.